test/raft/distributed_test.cc

```diff
@@ -613,5 +613,78 @@ struct failure_detector : public raft::failure_detector, public server_set {
     }
 };
 
+template <typename Payload>
+struct network {
+    struct message {
+        raft::server_id src;
+        raft::server_id dst;
+
+        // shared ptr to implement duplication of messages
+        lw_shared_ptr<Payload> payload;
+    };
+
+    struct event {
+        raft::logical_clock::time_point time;
+        message msg;
+    };
+
+    // A min-heap of event occurences compared by their time points.
+    std::vector<event> _events;
+    static bool cmp(const event& o1, const event& o2) {
+        return o1.time > o2.time;
+    }
+
+    raft::logical_clock _clock;
+
+    // When the time comes to deliver a message we use this function.
+    // However, based on the implementation, delivery may not be possible at the moment we try
+    // it (for example when delivery is implemented as pushing onto another queue which is currently
+    // full); this is represented by the returned boolean: delivery was successful if and only if
+    // the function returns true.
+    using deliver_t = std::function<bool(raft::server_id src, raft::server_id dst, const Payload&)>;
+    deliver_t _deliver;
+
+    // A pair (dst, [src1, src2, ...]) in this set denotes that `dst`
+    // does not receive messages from src1, src2, ...
+    std::unordered_map<raft::server_id, std::unordered_set<raft::server_id>> _grudges;
```

Comment:
Isn't this `std::unordered_multimap`?

Reply:
`unordered_multimap` doesn't support the operation I want to do (`_grudges[dst].contains(src)`).